src/storage/knn_index/knn_diskann/inner/diskann_utils.cppm

```diff
@@ -0,0 +1,481 @@
+// Copyright(C) 2023 InfiniFlow, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+module;
+
+
+#include <cstdlib>
+#include <boost/dynamic_bitset.hpp>
+
+import stl;
+import third_party;
+import default_values;
+import infinity_exception;
+import file_system;
+
+export module diskann_utils;
+
+namespace infinity {
+export void FileHandlerSeek(FileHandler &fh, i64 offset) {
+    auto &fs = fh.file_system_;
+    fs.Seek(fh, offset);
+}
+
+export inline void AllocAligned(void **ptr, SizeT size, SizeT align){
+    *ptr = nullptr;
+    if ((u64)size % (u64)align != 0){
+        UnrecoverableError(fmt::format("AllocAligned(): size {} is not aligned to {}", size, align));
+    }
+
+    *ptr = std::aligned_alloc(align, size);
+    if (*ptr == nullptr){
+        UnrecoverableError("AllocAligned(): Failed to allocate memory");
+    }
+}
+
+export inline void AlignedFree(void *ptr){
+    if (ptr == nullptr)
+        return;
+
+    free(ptr);
+}
+
+export struct Neighbor {
+    u32 id;
+    f32 distance;
+    bool expanded;
+
+    Neighbor() = default;
+
+    Neighbor(u32 id, f32 distance) : id(id), distance(distance), expanded(false) {}
+
+    inline bool operator<(const Neighbor &other) const {
+        return distance < other.distance || (distance == other.distance && id < other.id);
+    }
+
+    inline bool operator==(const Neighbor &other) const {
+        return (id == other.id);
+    }
+
+};
+
+export class NeighborPriorityQueue{
+public:
+    NeighborPriorityQueue(): size_(0), capacity_(0), cur_(0){}
+
+    NeighborPriorityQueue(SizeT capacity): size_(0), capacity_(capacity), cur_(0), data_{capacity + 1} {}
+
+    void Insert(const Neighbor &nbr) {
+        if (size_ == capacity_ && data_[size_ - 1] < nbr)
+            return;
+        
+        SizeT lo = 0, hi = size_;
+        while (lo < hi) { // binary search for insertion position
+            SizeT mid = (lo + hi) >> 1;
+            if (nbr < data_[mid])
+                hi = mid;
+            else if (data_[mid].id == nbr.id)
+                return;
+            else
+                lo = mid + 1;
+        }
+
+        if (lo < capacity_) {
+            std::memmove(&data_[lo + 1], &data_[lo], (size_ - lo) * sizeof(Neighbor));
```

Comment:
3th argument is (size_ - lo - 1) * sizeof(Neighbor). and data_ can have size capacity_

Reply:
(size_ - lo - 1) occurs error in boundary cases, and data_ initialises size to capacity+1 to avoid accessing undefined memory, so I didn't change it.